### ISSUE TITLE
[rb] - Adding/Editing Cookie issues (With suggestions)

### ISSUE THREAD

**Issue by luke-hill** [SeleniumHQ/selenium#8959]
## 🐛 Bug Report

Happy to fix these myself in 1/many PR's but the `#add_cookie` method needs some love. Specifically around the following.

- W3C Response of adding a cookie with expiration in past is 200 null. But cookie isn't created. **What should this be doing?**
- Error code when adding a cookie with Domain (Catch this earlier and throw nicer error in method)
- No `#edit_cookie`. Lets create an API that allows you to do so by caching the original cookie values, deleting the existing one then replacing it with a new one with the newly modified values
- `http_only` property not persisted and ignored in add method (This has been added by me recently in PR)
- `sameSite` property error (This might not be something we can control, **so possibly ignore**).

## To Reproduce

```

  cookie_manager.add_cookie(
    {
      :name=>"abcde",
      :value=>"3",
      :path=>"/",
      :domain=>Capybara.app_host, # This needs fixing / sanitizing out
      :expires=>DateTime.now.to_i, # This needs catching before sending (Or editing at w3c level)
      :same_site=>true,
      :secure=>true,
      :http_only=>true # This needs passing
    }
  )
```

## Expected behavior

Written in top

## Test script or set of commands reproducing this issue

As given

## Environment

Ruby Selenium 4alpha6 (Rest not relevant).

**Comment by luke-hill** [comment]
UPDATE:

If you attempt to add a cookie when it already exists, it acts as an updater
- Is this desirable?
- Is this logged anywhere. It feels it should be at least a WARN level message

You cannot ADD a cookie that is insecure
- Even when specifying secure false it does not work, it's being disregarded somewhere down the line (Possibly at bridge -> driver level. Because we pass the args in as symbols, maybe it doesn't like them?

**Comment by titusfortner** [comment]
Adding a cookie updates a cookie because w3c does not say to do anything different: https://w3c.github.io/webdriver/#add-cookie

There's nothing to log on Selenium side, it is a w3c / driver consideration

As for insecure cookies...
the "secure" parameter represents a `secure-only-flag` (https://w3c.github.io/webdriver/#cookies)
if you want it to be insecure, I believe you need to specify `httpOnly`

> Lets create an API that allows you to do so by caching the original cookie values, deleting the existing one then replacing it with a new one with the newly modified values

I don't think this is a good idea since Selenium doesn't control all the cookies in the browser, and we don't want to do a get Cookies command every time we add cookies. 

I think we updated same_site & http_only already? Is there more you want addressed here?

**Comment by luke-hill** [comment]
- Adding a cookie the past silently fails. Need something better.
- Cannot add a cookie using http only. This is where the API for getting / adding is different.

**Comment by titusfortner** [comment]
Ooh, good catch, the http only thing is a bug. Driver needs to see `httpOnly` and we are sending `http_only`
fixed: 2118f4a558326e644f4ab8b89f9fdd283ef74ebc

Setting a cookie to an expiration date in the past is another way of deleting an existing cookie. This is something that can be done by a browser, so Selenium shouldn't prevent you from doing it. This is more the kind of thing Capybara or Watir could implement. At the very least the behavior needs to be the same across all the languages.

**Comment by luke-hill** [comment]
Thanks for the explanation on the first point. I wasn't aware of this. I think possibly later we can look into the API and whether we could have some nice messages (even if they're just debug ones), saying things like, date is in the past warning this will delete the cookie.

As we discussed a while ago, I think this API across all languages needs standardising, the flipping between string/symbols here is a bit confusing too, but I think your patch looks like it'll do some initial good.
